File: src/java/com/kitware/pulse/cdm/patient/actions/SEPulmonaryShuntExacerbation.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package com.kitware.pulse.cdm.patient.actions;

import com.kitware.pulse.cdm.bind.PatientActions.PulmonaryShuntExacerbationData;
import com.kitware.pulse.cdm.properties.SEScalar0To1;

public class SEPulmonaryShuntExacerbation extends SEPatientAction
{

  private static final long serialVersionUID = 5044648250444265871L;
  protected SEScalar0To1 severity;
  
  public SEPulmonaryShuntExacerbation()
  {
    severity = null;
  }
  
  public void copy(SEPulmonaryShuntExacerbation other)
  {
    if(this==other)
      return;
    super.copy(other);
    if (other.severity != null)
      getSeverity().set(other.getSeverity());
    else if (severity != null)
      severity.invalidate();
  }
  
  @Override
  public void reset()
  {
    super.reset();
    if (severity != null)
      severity.invalidate();
  }
  
  @Override
  public boolean isValid()
  {
    return hasSeverity();
  }
  
  public static void load(PulmonaryShuntExacerbationData src, SEPulmonaryShuntExacerbation dst)
  {
    SEPatientAction.load(src.getPatientAction(), dst);
    if(src.hasSeverity())
      SEScalar0To1.load(src.getSeverity(),dst.getSeverity());
  }
  
  public static PulmonaryShuntExacerbationData unload(SEPulmonaryShuntExacerbation src)
  {
    PulmonaryShuntExacerbationData.Builder dst = PulmonaryShuntExacerbationData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  
  protected static void unload(SEPulmonaryShuntExacerbation src, PulmonaryShuntExacerbationData.Builder dst)
  {
    SEPatientAction.unload(src, dst.getPatientActionBuilder());
    if (src.hasSeverity())
      dst.setSeverity(SEScalar0To1.unload(src.severity));
  }
  

  public boolean hasSeverity()
  {
    return severity == null ? false : severity.isValid();
  }
  public SEScalar0To1 getSeverity()
  {
    if (severity == null)
      severity = new SEScalar0To1();
    return severity;
  }
  
  @Override
  public String toString()
  {
    return "Pulmonary Shunt Exacerbation" 
        + "\n\tSeverity: " + getSeverity();
  }
}

File: src/cpp/cdm/patient/actions/SEPulmonaryShuntExacerbation.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "patient/actions/SEPulmonaryShuntExacerbation.h"
#include "properties/SEScalar0To1.h"
#include "io/protobuf/PBPatientActions.h"

SEPulmonaryShuntExacerbation::SEPulmonaryShuntExacerbation(Logger* logger) : SEPatientAction(logger)
{
  m_Severity=nullptr;
}

SEPulmonaryShuntExacerbation::~SEPulmonaryShuntExacerbation()
{
  SAFE_DELETE(m_Severity);
}

void SEPulmonaryShuntExacerbation::Clear()
{
  SEPatientAction::Clear();
  INVALIDATE_PROPERTY(m_Severity);
}

void SEPulmonaryShuntExacerbation::Copy(const SEPulmonaryShuntExacerbation& src, bool preserveState)
{
  //if(preserveState) // Cache any state before copy,
  PBPatientAction::Copy(src, *this);
  //if(preserveState) // Put back any state
}

bool SEPulmonaryShuntExacerbation::IsValid() const
{
  return SEPatientAction::IsValid() && HasSeverity();
}

bool SEPulmonaryShuntExacerbation::IsActive() const
{
  if (!SEPatientAction::IsActive())
    return false;
  return !m_Severity->IsZero();
}
void SEPulmonaryShuntExacerbation::Deactivate()
{
  SEPatientAction::Deactivate();
  Clear();//No stateful properties
}

const SEScalar* SEPulmonaryShuntExacerbation::GetScalar(const std::string& name)
{
  if (name.compare("Severity") == 0)
    return &GetSeverity();
  return nullptr;
}

bool SEPulmonaryShuntExacerbation::HasSeverity() const
{
  return m_Severity==nullptr?false:m_Severity->IsValid();
}
SEScalar0To1& SEPulmonaryShuntExacerbation::GetSeverity()
{
  if(m_Severity==nullptr)
    m_Severity=new SEScalar0To1();
  return *m_Severity;
}
double SEPulmonaryShuntExacerbation::GetSeverity() const
{
  if (m_Severity == nullptr)
    return SEScalar::dNaN();
  return m_Severity->GetValue();
}

void SEPulmonaryShuntExacerbation::ToString(std::ostream &str) const
{
  str << "Patient Action : Pulmonary Shunt Exacerbation"; 
  if(HasComment())
    str<<"\n\tComment: "<<m_Comment;
  str << "\n\tSeverity: "; HasSeverity() ? str << *m_Severity : str << "Not Set";
  str << std::flush;
}